src/uct/ib/ud/base/ud_iface.c

```diff
@@ -310,12 +310,25 @@ ucs_status_t uct_ud_iface_complete_init(uct_ud_iface_t *iface)
         goto err;
     }
 
-    status = uct_ib_iface_event_fd_get(&iface->super.super.super, &event_fd);
-    if (status != UCS_OK) {
-        goto err_twheel_cleanup;
-    }
-
     if (iface->async.event_cb != NULL) {
+        uct_iface_attr_t iface_attr;
+        status = uct_ud_iface_query(iface, &iface_attr, 0, 0);
+        if (status != UCS_OK) {
+            goto err_twheel_cleanup;
+        }
+        if (!(iface_attr.cap.event_flags & (UCT_IFACE_FLAG_EVENT_FD |
+                                            UCT_IFACE_FLAG_EVENT_ASYNC_CB))) {
+            ucs_warn("Failed to set callback for async event handling. Device %s "
+                      "does not support any event notification mechanism",
+                       uct_ib_device_name(uct_ib_iface_device(&iface->super)));
+            return UCS_OK;
+        }
```

Comment:
why ud iface need to query itself?

Reply:
I can use `uct_ib_device_has_cq_notify()` instead.